tests/unit/s2n_client_hello_retry_test.c

```diff
@@ -480,6 +480,85 @@ int main(int argc, char **argv)
+
+        EXPECT_NOT_NULL(server_conn = s2n_connection_new(S2N_SERVER));
+        EXPECT_NOT_NULL(client_conn = s2n_connection_new(S2N_CLIENT));
+
+        EXPECT_NOT_NULL(tls13_chain_and_key = s2n_cert_chain_and_key_new());
+        EXPECT_SUCCESS(s2n_read_test_pem(S2N_ECDSA_P384_PKCS1_CERT_CHAIN, tls13_cert_chain, S2N_MAX_TEST_PEM_SIZE));
+        EXPECT_SUCCESS(s2n_read_test_pem(S2N_ECDSA_P384_PKCS1_KEY, tls13_private_key, S2N_MAX_TEST_PEM_SIZE));
+        EXPECT_SUCCESS(s2n_cert_chain_and_key_load_pem(tls13_chain_and_key, tls13_cert_chain, tls13_private_key));
+        EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(client_config, tls13_chain_and_key));
+        EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(server_config, tls13_chain_and_key));
+
+
+        EXPECT_SUCCESS(s2n_connection_set_config(server_conn, server_config));
+
+        /* The client will offer the default tls13 ciphersuites */
+        s2n_connection_set_cipher_preferences(client_conn, "default_tls13");
+
+        /* Force the client to send an empty list of keyshares */
+        EXPECT_SUCCESS(s2n_connection_set_keyshare_by_name_for_testing(client_conn, "none"));
+
+        /* ClientHello 1 */
+        EXPECT_SUCCESS(s2n_client_hello_send(client_conn));
+        EXPECT_SUCCESS(s2n_stuffer_copy(&client_conn->handshake.io, &server_conn->handshake.io,
+                                        s2n_stuffer_data_available(&client_conn->handshake.io)));
+        EXPECT_SUCCESS(s2n_client_hello_recv(server_conn));
+        EXPECT_TRUE(s2n_is_hello_retry_handshake(server_conn));
+        EXPECT_SUCCESS(s2n_set_connection_hello_retry_flags(server_conn));
+
+        /* Server HelloRetryRequest */
+        EXPECT_SUCCESS(s2n_server_hello_retry_send(server_conn));
+        EXPECT_SUCCESS(s2n_set_connection_hello_retry_flags(server_conn));
+
+        EXPECT_SUCCESS(s2n_stuffer_wipe(&client_conn->handshake.io));
+        EXPECT_SUCCESS(s2n_stuffer_copy(&server_conn->handshake.io, &client_conn->handshake.io,
+                                        s2n_stuffer_data_available(&server_conn->handshake.io)));
+        client_conn->handshake.message_number = HELLO_RETRY_MSG_NO;
+        EXPECT_SUCCESS(s2n_server_hello_recv(client_conn));
+
+        EXPECT_EQUAL(client_conn->secure.cipher_suite, &s2n_tls13_aes_256_gcm_sha384);
+
+        /* ClientHello 2 */
+        EXPECT_SUCCESS(s2n_client_hello_send(client_conn));
+        EXPECT_SUCCESS(s2n_stuffer_wipe(&server_conn->handshake.io));
+        EXPECT_SUCCESS(s2n_stuffer_copy(&client_conn->handshake.io, &server_conn->handshake.io,
+                                        s2n_stuffer_data_available(&client_conn->handshake.io)));
+        EXPECT_SUCCESS(s2n_client_hello_recv(server_conn));
+
+        /* Server Hello with cipher suite that does not match Hello Retry Request cipher suite */
+        server_conn->secure.cipher_suite = &s2n_tls13_chacha20_poly1305_sha256;
```

Comment:
If you're just going to set the secure.cipher_suite manually, what is the point of sending/receiving the client_hellos and retry? Can we skip that?

Reply:
Yeah good point, we don't need the client hello/hrr calls.